### PR TITLE
Mmahsereci/qrbf for uniform measure

### PR DESCRIPTION
**Pull request by mmahsereci**
*Issue #, if available:*

*Description of changes:*
Adding RBF integrability against Uniform measure, both for finite and infinite bounds. I did not address the re-structuring of the integral test yet. That'll be another PR.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/quadrature/kernels/quadrature_rbf.py [hunk 1]
```diff
@@ -218,3 +219,88 @@ def dqK_dx(self, x2: np.ndarray) -> np.ndarray:
...
+        """
+        lower_bounds = np.array([b[0] for b in self._bounds_list_for_kernel_integrals])
+        upper_bounds = np.array([b[1] for b in self._bounds_list_for_kernel_integrals])
+        erf_lo = erf(self._scaled_vector_diff(lower_bounds, x2))
+        erf_up = erf(self._scaled_vector_diff(upper_bounds, x2))
+        kernel_mean = self.variance * (self.lengthscale * np.sqrt(np.pi / 2.) * (erf_up - erf_lo)).prod(axis=1)
+
+        return kernel_mean.reshape(1, -1) * self.measure.density
+
+    def qKq(self) -> np.float:
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply emukit/quadrature/kernels/quadrature_rbf.py]: done

File: emukit/model_wrappers/gpy_quadrature_wrappers.py [hunk 2]
```diff
@@ -199,13 +199,17 @@ def create_emukit_model_from_gpy_model(gpy_model: GPy.models.GPRegression,
     # infinite bounds. Gauss yes, uniform measure not implemented yet
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply emukit/model_wrappers/gpy_quadrature_wrappers.py]: yes, I updated it.

File: emukit/model_wrappers/gpy_quadrature_wrappers.py [hunk 3]
```diff
@@ -216,7 +220,10 @@ def create_emukit_model_from_gpy_model(gpy_model: GPy.models.GPRegression,
     # finite bounds: measure: uniform only not implemented
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply emukit/model_wrappers/gpy_quadrature_wrappers.py]: thanks, I updated it

File: emukit/quadrature/kernels/quadrature_rbf.py [hunk 4]
```diff
@@ -218,3 +219,88 @@ def dqK_dx(self, x2: np.ndarray) -> np.ndarray:
...
+class QuadratureRBFUniformMeasure(QuadratureRBF):
+    """
+    And RBF kernel with integrability over a uniform measure. Can be used with finite as well as infinite integral
+    bounds.
+
+    Note that each standard kernel goes with a corresponding quadrature kernel, in this case standard rbf kernel.
+    """
+
+    def __init__(self, rbf_kernel: IRBF, integral_bounds: Optional[List[Tuple[float, float]]],
+                 measure=UniformMeasure, variable_names: str=''):
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply emukit/quadrature/kernels/quadrature_rbf.py]: done

File: emukit/quadrature/kernels/quadrature_rbf.py [hunk 5]
```diff
@@ -218,3 +219,88 @@ def dqK_dx(self, x2: np.ndarray) -> np.ndarray:
...
+        for (lb_d, ub_d) in bounds:
+            if lb_d >= ub_d:
+                raise ValueError("Upper bound of relevant integration domain must be larger than lower bound. Found a "
+                                 "pair containing ({}, {}).".format(lb_d, ub_d))
+        self._bounds_list_for_kernel_integrals = bounds
+        self.reasonable_box_bounds = BoxBounds(name=variable_names, bounds=bounds)
+
+    def qK(self, x2: np.ndarray) -> np.ndarray:
+        """
+        RBF kernel with the first component integrated out aka. kernel mean
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply emukit/quadrature/kernels/quadrature_rbf.py]: done

File: tests/emukit/quadrature/ground_truth_integrals_qkernel.py [hunk 6]
```diff
@@ -24,12 +24,13 @@ def _sample_uniform(num_samples: int, bounds: List[Tuple[float, float]]):
...
 def _sample_gauss_iso(num_samples: int, measure: IsotropicGaussianMeasure):
-    samples = np.reshape(np.random.randn(num_samples * measure.dim), [num_samples, D])
+    D = measure.dim
+    samples = np.reshape(np.random.randn(num_samples * D), [num_samples, D])
     return measure.mean + np.sqrt(measure.variance) * samples
 
 
-# qK integrals
-def qK_lebesgue_measure(num_samples: int, qrbf: QuadratureRBFLebesgueMeasure, x2: np.ndarray):
+# lebesque
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply tests/emukit/quadrature/ground_truth_integrals_qkernel.py]: removed

File: tests/emukit/quadrature/ground_truth_integrals_qkernel.py [hunk 7]
```diff
@@ -54,41 +48,85 @@ def qKq_lebesgue_measure(num_samples: int, qrbf: QuadratureRBFLebesgueMeasure):
     return np.mean(qKx) * volume
 
 
-def qKq_gauss_iso(num_samples: int, measure: IsotropicGaussianMeasure, qrbf: QuadratureRBFIsoGaussMeasure):
+# Gauss iso
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply tests/emukit/quadrature/ground_truth_integrals_qkernel.py]: removed

File: tests/emukit/quadrature/ground_truth_integrals_qkernel.py [hunk 8]
```diff
@@ -54,41 +48,85 @@ def qKq_lebesgue_measure(num_samples: int, qrbf: QuadratureRBFLebesgueMeasure):
...
+
+
+def qKq_gauss_iso(num_samples: int, qrbf: QuadratureRBFIsoGaussMeasure):
+    measure = qrbf.measure
     samples = _sample_gauss_iso(num_samples, measure)
     qKx = qrbf.qK(samples)
     return np.mean(qKx)
 
 
+# uniform
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply tests/emukit/quadrature/ground_truth_integrals_qkernel.py]: removed

File: tests/emukit/quadrature/ground_truth_integrals_qkernel.py [hunk 9]
```diff
@@ -54,41 +48,85 @@ def qKq_lebesgue_measure(num_samples: int, qrbf: QuadratureRBFLebesgueMeasure):
...
     np.random.seed(0)
 
     # === Choose MEASURE BELOW ======
-    MEASURE = 'Lebesgue'
-    #MEASURE = 'GaussIso'
+    #MEASURE_INTBOUNDS = 'Lebesgue-finite'
+    #MEASURE_INTBOUNDS = 'GaussIso-infinite'
+    #MEASURE_INTBOUNDS = 'Uniform-infinite'
+    MEASURE_INTBOUNDS = 'Uniform-finite'
     # === CHOOSE MEASURE ABOVE ======
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply tests/emukit/quadrature/ground_truth_integrals_qkernel.py]: thanks, this is in the script that generates the ground truth integrals via sampling. I'll change this when I re-do the tests as discussed, OK?

[the remark this answers is not in the document]
Comment (apaleyes) [reply tests/emukit/quadrature/ground_truth_integrals_qkernel.py]: ok!

**Pull request merged**